File: src/java1/algorithms/array/threeSumDuplicates/ThreeSumDuplicates.java
package threeSumDuplicates;
import java.util.*;

// Hashing:- TC: O(n logn) + O(n*2) => O(n * 2), SC: O(n)
public class ThreeSumDuplicates {
    private static List<List<Integer>> threeSum(int[] nums, int target) {
        List<List<Integer>> triplets = new ArrayList<>();
        HashSet<Integer> hashset = new HashSet<>();

        for(int i=0; i< nums.length-2; i++) {
            for(int j=i+1; j< nums.length-1; j++) {
                int complement = target -(nums[i]+nums[j]);
                if(hashset.contains(complement) && !triplets.contains(List.of(nums[i], nums[j], complement))) {
                    triplets.add(List.of(nums[i], nums[j], complement));
                } else {
                    hashset.add(nums[j]);
                }
            }
        }
        return triplets;
    }

    public static void main(String[] args) {
        int[] nums = {3, 3, -1, -2, 0, -2, 0, -1, 0, -1};
        System.out.println(threeSum(nums, 0));
    }
}

File: src/javascript/algorithms/array/threeSumDuplicates/threeSumDuplicates.js
// Hashing:- TC: O(n logn) + O(n^2) => O(n^2), SC: O(n)
function threeSumHashing(nums, target) {
    let triplets = [];
    let tripletSet = new Set();
    for(let i =0; i < nums.length - 2; i++) {
        for(let j=i+1; j< nums.length - 1; j++) {
            let complement = target - (nums[i] + nums[j]);
            if(tripletSet.has(complement)) {
                triplets.push([nums[i], nums[j], complement]);
            } else {
                tripletSet.add(nums[j]);
            }
        }
    }
    return triplets;
}

let nums1 = [1, 2, 3, 4, 5, 6, 7, 8];
let nums2 = [-1,0,1,2,-1,-4];
console.log(threeSumHashing(nums1, 9));
console.log(threeSumHashing(nums2, 0));